Introduce IWritableStorage interface

nupplaphil opened the pull request:
This PR is another preparation for additional, selectable storages.

I extracted all methods for selectable storage backends (currently Filesystem & DB) from `IStorage` to `ISelectableStorage`. So all the `BadMethodCallException`s aren't necessary anymore. I split the corresponding `get` operations of the StorageManager as well. So there's no need for an extra parameter `userBackend` anymore. Instead you use either the `getByName()` method to geht ANY backend, or you use the `getSelectableByName()` to get a specific, selectable backend. No need for any other checks.

Another improvement is the error handling. The backend per se doesn't need to know what the caller tries to do with the content. Therefore it shouldn't decide if an error of the persistence-layer is "acceptable" or not, it just throws an exception in case the call isn't right (e.g. you want to delete something, which doesn't exist, or get something, which doesn't exist). The caller is now responsible to catch the exception and do whatever he needs to do :).

Therefore I created two exceptions, the `ReferenceStorageExecption` for exceptions, where the caller uses invalid references and the `StorageException` for "the rest".

I used this branch (with Filesystem backend) for a couple of days without any exception so far.

nupplaphil committed:
Split IStorage and ISelectableStorage and make their behaviour homogenous

nupplaphil committed:
Refactor IStorage

nupplaphil committed:
Fix Storage Exceptions

nupplaphil committed:
Fixup Database behaviour

nupplaphil committed:
Update AddonStorageBackend.md

annando commented:
Would be great having a WebDAV storage.

MrPetovan commented on doc/AddonStorageBackend.md:
```diff
@@ -106,21 +106,56 @@ Add a new test class which's naming convention is `StorageClassTest`, which exte
+
+### `StorageException`
+
+This is the common exception in case during the storage backend occurs any, unexpected errors.
```
```suggestion
This is the common exception in case unexpected errors happen using the storage backend.
```

MrPetovan commented on doc/AddonStorageBackend.md:
```diff
@@ -106,21 +106,56 @@ Add a new test class which's naming convention is `StorageClassTest`, which exte
+### `StorageException`
+
+This is the common exception in case during the storage backend occurs any, unexpected errors.
+If there's a predecessor to this execption (e.g. you catched a exception an re-throw this execption), you should add the predecessor for transparent reasons.
```
```suggestion
If there's a predecessor to this exception (e.g. you caught an exception and are throwing this execption), you should add the predecessor for transparency reasons.
```

MrPetovan commented on doc/AddonStorageBackend.md:
````diff
@@ -106,21 +106,56 @@ Add a new test class which's naming convention is `StorageClassTest`, which exte
+} 
+```
+
+### `ReferenceStorageExecption`
````
Please move this more specific exception first.

nupplaphil replied:
done

MrPetovan commented on doc/AddonStorageBackend.md:
```diff
@@ -106,21 +106,56 @@ Add a new test class which's naming convention is `StorageClassTest`, which exte
 ## Example
 
-Here an hypotetical addon which register an unusefull storage backend.
+Here an hypotetical addon which register an unuseful storage backend.
```
```suggestion
Here an hypotetical addon which register a useless storage backend.
```

annando commented:
From my part it looks okay, but due to the complexity it would be great if @MrPetovan had got a look at it as well.

MrPetovan commented:
I started looking at it in a car yesterday, but it was too big for my phone.

MrPetovan commented on src/Model/Photo.php:
```diff
@@ -198,12 +201,17 @@ public static function getImageDataForPhoto(array $photo)
 			$i = self::selectFirst(['data'], ['id' => $photo['id']]);
 			if ($i === false) {
-				return null;
+				return '';
```
Why the change of return value type?

nupplaphil replied:
I think it's out of scope as well, but the callers are using the return value either as object .. or as string .. see:
https://github.com/friendica/friendica/blob/c6b929d6d932f1881e0697356efa702f00e2981b/src/Model/Photo.php#L230-L235
The first argument of `Image` assert that it's a string, not an object

I'll revert it, because I'm not sure if there're side effects ..

MrPetovan commented on src/Core/StorageManager.php:
```diff
@@ -73,65 +71,111 @@ class StorageManager
+	 *
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 */
+	public function getSelectableStorageByName(string $name = null)
```
Shouldn't this method throw an exception when the backend isn't valid?

nupplaphil replied:
Done

nupplaphil replied:
I reverted this change, because there are a lot of places (mainly inside the legacy api), where an "invalid" backend results in using the legacy storage method.

MrPetovan replied:
Can't this be done in a try-catch block instead of checking the returned value of the method?

nupplaphil replied:
You're right .. but I need to adapt all calls so I tried to avoid it :D

MrPetovan replied:
When a function main use is this:
```php
$return = function($param);
if ($return) {
   ...
} else {
   // Awkward error handling
}
```
Then it really should be
```php
try {
    $return = function($param);
    ...
} catch (Throwable $e) {
    // Proper error handling
}
```
The structures are similar and easy to transform from one to the other, it just is more explicit in the latter case.

nupplaphil replied:
of course, you're right, but I need to take into account that in case of the "legacy" backend, there's no "current backend" set, so it would throw this exception all the time when calling `DI::storage()` ..

MrPetovan commented on src/Model/Storage/Database.php:
```diff
@@ -39,47 +38,57 @@ class Database extends AbstractStorage
-	 * @param LoggerInterface $logger
-	 * @param L10n            $l10n
 	 */
-	public function __construct(DBA $dba, LoggerInterface $logger, L10n $l10n)
```
Why did you forgo the logger and the l10n dependencies?

nupplaphil replied:
I thought the exceptions superseded the log-points. For me there's no need to call a log-point and throw the same exception to the caller.
l10n --> see https://github.com/friendica/friendica/pull/10582#discussion_r686248259

MrPetovan commented on src/Model/Storage/Filesystem.php:
```diff
@@ -96,8 +97,7 @@ private function createFoldersForFile(string $file)
 			if (!mkdir($path, 0770, true)) {
-				$this->logger->warning('Failed to create dir.', ['path' => $path]);
-				throw new StorageException($this->l10n->t('Filesystem storage failed to create "%s". Check you write permissions.', $path));
+				throw new StorageException(sprintf('Filesystem storage failed to create "%s". Check you write permissions.', $path));
```
Why didn't you keep the localized error messages?

nupplaphil replied:
There's no need for localized errors at backend exceptions. They aren't meant to be printed at the frontend, so it could happen that there's a catch with a log-command and so you would get a (user-based) translated exception message inside your log :D

MrPetovan replied:
It could happen, but does it happen?

nupplaphil replied:
OK, but it's one of the rare places where we use translated exceptions .. But I don't see a use case for it .. what's the reason for translated exceptions? For me it just feels wrong tbh ^^ .. exceptions are "unwanted behaviors", mostly because of data/logic issues, addressed to admins and developers

MrPetovan replied:
But if it's triggered by a user, and we transmit the exception message to display, then it's important to show it in their preferred language. I understand that it wouldn't be of much help to them, but it still is more welcoming than error messages in a language you don't master.

nupplaphil replied:
currently we wouldn't display the exception to the user. There's a catch around it and a log-point *gg*. see:
https://github.com/friendica/friendica/blob/0c9163f299e693c853cbcecd05b88523ec9a1a3f/src/Model/Photo.php#L193-L217

But this behavior is now more straight, because before this change, the DB call returned null and the storage call threws an exception. I'll add a generic check at the `Module\Photo` and display the output to the user.

MrPetovan replied:
This works for me.

MrPetovan commented on src/Model/Storage/ISelectableStorage.php:
```diff
@@ -0,0 +1,100 @@
+namespace Friendica\Model\Storage;
+
+/**
+ * Interface for selectable storage backends
```
Can you please extend on what "selectable" means in this context?

nupplaphil replied:
Maybe you have a better naming for this interface: 

It's a storage backend, which an admin can select for "user data" and have the CRUD (create update delete) pattern. The `IStorage` is a much simpler interface which allows just to retrieve data from a storage backend.

MrPetovan replied:
Why not `Writable`?

nupplaphil replied:
sounds better ^^

MrPetovan commented on src/Model/Storage/ReferenceStorageException.php:
```diff
@@ -21,31 +21,11 @@
+ * Storage Exception in case of invalid references
  */
-abstract class AbstractStorage implements IStorage
+class ReferenceStorageException extends Exception
```
Shouldn't it extend `StorageException`?

nupplaphil replied:
hmm ... good catch .. I wasn't sure if the `RefernceStorageException` is a child or a sibling ... but you're right, I guess child is better ...

nupplaphil replied:
Done

MrPetovan commented on src/Model/Storage/ReferenceStorageException.php:
```diff
@@ -21,31 +21,11 @@
 
 namespace Friendica\Model\Storage;
```
Not sure these exception classes should be in the `Model\` namespace, maybe we should start a new namespace for all the Friendica exceptions.

nupplaphil replied:
I'd prefer having the exceptions at the place where the business logic resides. I think it's better to have explicit exceptions rather than "global" exceptions. f.e. I'm not a fan that we use the `InternalServerException` in cases of the backend errors. It's too common and should be the "Response" to a caller, not the "root-cause"

MrPetovan replied:
Got it.

MrPetovan commented on src/Util/HTTPSignature.php:
```diff
@@ -408,7 +409,7 @@ public static function fetch($request, $uid)
 	 *                         'cookiejar' => path to cookie jar file
 	 *
-	 * @return object CurlResult
+	 * @return CurlResult|void CurlResult
```
This shouldn't return `void` but throw an Exception if the `$owner` variable is empty, but maybe it's outside of the scope of this PR.

nupplaphil replied:
I reverted it .. it's really part of another PR

The reason for it: https://github.com/friendica/friendica/blob/218d451248c585b9c97915d852d4bacbff838ecf/src/Util/HTTPSignature.php#L422 and https://github.com/friendica/friendica/blob/218d451248c585b9c97915d852d4bacbff838ecf/src/Util/HTTPSignature.php#L427

annando commented:
@nupplaphil have a look at the conflict.

nupplaphil committed:
- Moved the description for the specific storage exception first
- Introduced exceptions for try to get invalid storage
- ReferenceStorageException now extends StorageException

nupplaphil committed:
- Revert HTTPSignature change

nupplaphil committed:
Revert Photo::getImageDataForPhoto return-type change

nupplaphil committed:
Rename ISelectableStorage to IWritableStorage

nupplaphil committed:
Force "Database" as default storage backend

nupplaphil committed:
Introduce InvalidClassStorageException and adapt the code for it

nupplaphil commented:
OK, this is now A LOT bigger than I wanted .. I'm testing the current branch for a couple of days on my node ..

MrPetovan commented on mod/photos.php:
```diff
@@ -323,7 +323,11 @@ function photos_post(App $a)
+					$image = Photo::getImageForPhoto($photo);
+				} catch (Exception $exception) {
+					throw new HTTPException\InternalServerErrorException(DI::l10n()->t('Internal server error'), $exception);
+				}
```
The `Image` class already has a built-in method to check the validity of the parameter, please add this try-catch block in `Model\Photo::getImageDataForPhoto()` and make `Model\Photo::getImageForPhoto()` return an `Object\Image` instance no matter what.

This saves you having to add this try-catch block everywhere since we were already using `Object\Image->isValid()`.

MrPetovan commented on src/Core/StorageManager.php:
```diff
@@ -72,82 +69,135 @@ class StorageManager
+	 * @throws Storage\StorageException in case of an unexpected failure during the hook call
 	 */
-	public function getByName(string $name = null, $onlyUserBackend = false)
+	public function getWritableStorageByName(string $name)
```
This is madness, you have two identical methods except one method call parameter changed from `true` to `false`. A better way to do this is to pass the array of valid backends to `getByName` like this:
```php
public function getWritableStorageByName(string $name)
{
    return $this->getByName($name, $this->backends);
}

public function getByName(string $name, $validBackends = null)
{
    ...
    if ($this->isValidBackend($name, $validBackends)) {
        ...
    }
    ...
}

public function isValidBackend(string $name = null, array $validBackends = null): bool
{
    $validBackends = $validBackends ?? array_merge($this->backends, [Storage\SystemResource::getName(), Storage\ExternalResource::getName()]);
    return in_array($name, $validBackends);
}
```

nupplaphil replied:
Looks a lot more readable, thx for this hint!

MrPetovan commented on src/Model/Photo.php:
```diff
@@ -186,26 +189,29 @@ public static function exists(array $conditions)
+		try {
+			$backendClass = DI::storageManager()->getByName($photo['backend-class'] ?? '');
+			$backendRef   = $photo['backend-ref'] ?? '';
+			return $backendClass->get($backendRef);
```
Unnecessary intermediate variable.
```suggestion
			return $backendClass->get($photo['backend-ref'] ?? '');
```

annando commented on database.sql:
```diff
@@ -1,6 +1,6 @@
 -- ------------------------------------------
 -- Friendica 2021.09-dev (Siberian Iris)
--- DB_UPDATE_VERSION 1432
+-- DB_UPDATE_VERSION 1433
```
If my current PR #10593 is merged first, you will have to manually increate the database version. The system will not show a conflict here.

nupplaphil replied:
addressed here: https://github.com/friendica/friendica/pull/10582/commits/92d532db6cac4a52840a003a3e62e3a4064bed51
thx for the hint!

MrPetovan commented on src/Model/Storage/Filesystem.php:
```diff
@@ -118,23 +118,34 @@ private function createFoldersForFile(string $file)
+
+		$result = file_get_contents($file);
+
+		// just in case the result is REALLY false, not zero or empty or anything else, throw the exception
```
This comment is entirely redundant with the code itself.
```suggestion
```

MrPetovan commented on src/Model/Storage/IWritableStorage.php:
```diff
@@ -0,0 +1,103 @@
+ * Interface for writable storage backends
+ *
+ * Used for storages with CRUD functionality, mainly used for user data (e.g. photos, attachements).
+ * There's only one active, writable storage possible. This type of storages are selectable by the current administrator
```
Did you have a free comma you had to put somewhere? 😄
```suggestion
 * There's only one active writable storage possible. This type of storage is selectable by the current administrator.
```

nupplaphil committed:
Update src/Model/Storage/IWritableStorage.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

nupplaphil committed:
Update src/Model/Storage/Filesystem.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

nupplaphil committed:
Update db version

nupplaphil committed:
Add todo for later

nupplaphil commented:
@MrPetovan I think I'm done , I will watch the error output until tomorrow and mark it ready for merge, but you already can take a look once more

MrPetovan commented on src/Core/StorageManager.php:
```diff
@@ -46,7 +45,7 @@ class StorageManager
 	];
 
-	private $backends = [];
+	private $backends;
```
We should explicit this array as a list of IWritableStorage class names.

nupplaphil committed:
Lower complexity for valid backends (replace hashmap with a "simple" name array)

nupplaphil committed:
Fix usage

MrPetovan reviewed (approved):
I asked for a one-line comment 😂

nupplaphil commented:
seems to work .. ready for merge :)

MrPetovan commented on src/Core/StorageManager.php:
```diff
@@ -61,93 +62,115 @@ class StorageManager
+			return $storage;
+		} else {
+			throw new Storage\InvalidClassStorageException(sprintf('Backend %s is not writable', $name));
+		}
```
Nitpicking, this structure can get rid of the `else`.
```suggestion
		if (!$storage instanceof Storage\IWritableStorage) {
			throw new Storage\InvalidClassStorageException(sprintf('Backend %s is not writable', $name));
		}
		
		return $storage;
```

MrPetovan commented on src/Core/StorageManager.php:
```diff
@@ -61,93 +62,115 @@ class StorageManager
 			// If the current name isn't a valid backend (or the SystemResource instance) create it
-			if ($this->isValidBackend($name, $onlyUserBackend)) {
+			if ($this->isValidBackend($name, $validBackends)) {
 				switch ($name) {
```
Same principle here with removing an indentation level.
```suggestion
			if (!$this->isValidBackend($name, $validBackends)) {
				throw new Storage\InvalidClassStorageException(sprintf('Backend %s is not valid', $name));
			}
			
			switch ($name) {
```

MrPetovan commented on src/Core/StorageManager.php:
```diff
@@ -61,93 +62,115 @@ class StorageManager
+								$this->backendInstances[$data['name'] ?? $name] = $data['storage'];
+							} else {
+								throw new Storage\InvalidClassStorageException(sprintf('Backend %s was not found', $name));
+							}
```
More simplification
```suggestion
							if (!($data['storage'] ?? null) instanceof Storage\IStorage) {
								throw new Storage\InvalidClassStorageException(sprintf('Backend %s was not found', $name));
							}
							
							$this->backendInstances[$data['name'] ?? $name] = $data['storage'];
```

MrPetovan commented on src/Core/StorageManager.php:
```diff
@@ -227,16 +231,20 @@ public function listBackends()
 
-			$backends                    = $this->backends;
-			$backends[$class::getName()] = $class;
+			if (array_search($class::getName(), $this->validBackends, true) !== false) {
```
Look at this!
```suggestion
			if ($this->isValidBackend($class::getName(), $this->validBackends)) {
```

nupplaphil committed:
Update src/Core/StorageManager.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

MrPetovan commented on src/Core/StorageManager.php:
```diff
@@ -61,93 +62,118 @@ class StorageManager
+				throw new Storage\InvalidClassStorageException(sprintf('Backend %s is not valid', $name));
+			}
+			
+			switch ($name) {
```
I wasn't able to suggest the full modification of the file, you have to deindent the rest of the `switch` cases and remove the final `else` yourself.

nupplaphil replied:
I already saw it :-) .. I rebased the current branch onto the develop branch and did some "smoke tests" again (upload image, post image, set image as avatar) ... thx for the hint!

nupplaphil committed:
Update messages.po
